### PR TITLE
[NNPA] Split a big matmul for NNPA

### PR DESCRIPTION
**Pull request by tungld**
This patch adds a pattern to split a large MatMul into smaller ones that fit into NNPA. Given `NxK * K*M`, the pattern considers dimensions `N` and/or `M` to split. If `N` and/or `M` is greater than NNPA_MAXIMUM_DIMENSION_INDEX_SIZE.

For example, `NxK * KxM` will be rewritten into `(N1xK * KxM1 ++ N1xK * KxM2) ++ (N2xK * KxM1 ++ N2xK * KxM2)`, where `N = N1 ++ N2`, `M = M1 ++ M2`, and `++` is concatenation along an appropriate dimension.

Tensors are splitted into chunks of the equal size of NNPA_MAXIMUM_DIMENSION_INDEX_SIZE, except the last chunk.

### CODE REVIEW

File: src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp [hunk 1]
```diff
@@ -281,6 +282,67 @@ Type CreatePaddedXType(Value x, ArrayAttr pads) {
   return paddedType;
 }
 
+class SplitLargeMatMulPattern : public OpRewritePattern<ONNXMatMulOp> {
```
Comment (AlexandreEichenberger) [inline src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp:285]: I know it's early in development, please add a comment to explain the pattern using math mul tensors, that would be helpful for current and further reading of the code, tx

Comment (tungld) [reply src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp:296]: Added a comment. Thanks!

**Comment by tungld** [comment]
> Do we need to pass to split all ZHigh op if their size is too large? Will this data decomposition be related to data level parallelization?

Do you mean a separate pass for splitting? I think it would be a good idea.

> Will this data decomposition be related to data level parallelization?

Yes, somewhat related. However, sub-matmul ops still runs in a sequential way on NNPA.

File: src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp [hunk 2]
```diff
@@ -372,18 +463,24 @@ void RewriteONNXForZHighPass::runOnOperation() {
...
 
-    // - one input is N-D, N > 3 and the other is 2-D.
+    // - one input is N-D (N > 3) and the other is 2-D.
     if (aRank == 2 && bRank > 3)
       return false;
     if (bRank == 2 && aRank > 3)
       return false;
+    // No input is N-D (N > 3) but dimension N or M (NxK * KxM) exceeds NNPA
+    // limitation.
+    if ((aRank == 3 || aRank == 3) && (bRank == 2 || bRank == 3) &&
```
[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp:476]: Should you not also tests the K dim, and the broadcasts

File: src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp [hunk 3]
```diff
@@ -281,6 +282,94 @@ Type CreatePaddedXType(Value x, ArrayAttr pads) {
...
 
+/// This pattern is to split a large MatMul into smaller ones that fit into
+/// NNPA. Given (NxK) * (K*M), the pattern considers dimensions N and/or M to
+/// split, if N and/or M is greater than NNPA_MAXIMUM_DIMENSION_INDEX_SIZE.
+/// For example,
+///   (NxK) * (KxM) will be rewritten into
+///   (N1xK * KxM1 ++ N1xK * KxM2) ++ (N2xK * KxM1 ++ N2xK * KxM2)
+/// where N = N1 ++ N2, M = M1 ++ M2, and `++` is concatenation along an
+/// appropriate dimension.
+///
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp:291]: Now I added a figure to explain how to split matrices.

This patch only splits N and M dimensions, it does not check K and batch dimensions. As you said, yes, it's possible to deal with K and batch dimensions. Let me do it later.

File: src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp [hunk 4]
```diff
@@ -281,6 +282,94 @@ Type CreatePaddedXType(Value x, ArrayAttr pads) {
...
+    }
+    Value res = resSubAs[0];
+    if (resSubAs.size() > 1)
+      // Concat sub results along dimension N of A.
+      res = create.onnx.concat(outputType, resSubAs, outputRank - 2);
+
+    rewriter.replaceOp(op, res);
+    return success();
+  }
+};
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp:389]: I see. Thanks!

File: src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp [hunk 5]
```diff
@@ -372,18 +463,24 @@ void RewriteONNXForZHighPass::runOnOperation() {
...
-    // - one input is N-D, N > 3 and the other is 2-D.
+    // - one input is N-D (N > 3) and the other is 2-D.
     if (aRank == 2 && bRank > 3)
       return false;
     if (bRank == 2 && aRank > 3)
       return false;
+    // No input is N-D (N > 3) but dimension N or M (NxK * KxM) exceeds NNPA
+    // limitation.
+    if ((aRank == 3 || aRank == 3) && (bRank == 2 || bRank == 3) &&
+        ((aShape[aRank - 2] > NNPA_MAXIMUM_DIMENSION_INDEX_SIZE) ||
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp:496]: Yes, explained this in the comment now.

File: src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp [hunk 6]
```diff
@@ -360,10 +449,12 @@ void RewriteONNXForZHighPass::runOnOperation() {
   });
 
   // Illegalize MatMulOp if
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp]: I see. I updated the comment.

File: src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (tungld) [reply src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp]: Oops, fixed it. Thanks!

**Pull request merged**